### PR TITLE
chore: introduce sorted_map

### PR DESCRIPTION
**Pull request by romange**
Consolidate skiplist based zset functionality into a dedicated class.

### CODE REVIEW

File: src/core/zset_test.cc [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#include "core/mi_memory_resource.h"
+
+extern "C" {
+#include "redis/zmalloc.h"
+#include "redis/zset.h"
+}
+
+namespace dfly {
+
+class ZSetTest : public ::testing::Test {
```
Comment (romange) [inline src/core/zset_test.cc:19]: this unit test is not directly related to this Pr but it shows the potential of using a different data structure than skiplist. In this case I compared it to absl::btree.

File: src/core/sorted_map.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+class SortedMap {
+ public:
+  using ScoredMember = std::pair<std::string, double>;
+  using ScoredArray = std::vector<ScoredMember>;
+
+  SortedMap();
+  ~SortedMap();
+
+  static SortedMap* FromListPack(const uint8_t* lp);
```
[the remark this answers is not in the document]
Comment (romange) [reply src/core/sorted_map.h:28]: The reason I return a pointer is because the context demands so. I have no use for this object on the stack because I must pass its pointer into the CompactObject.

Please note that while it looks like a new class, it's really about consolidation of dispersed code under a single entity. 
I will add an ownership comment.

File: src/core/sorted_map.cc [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+int SortedMap::Add(double score, sds ele, int in_flags, int* out_flags, double* newscore) {
+  zskiplistNode* znode;
+  dictEntry* de = dictFind(dict_, ele);
+
+  /* Turn options into simple to check vars. */
+  int incr = (in_flags & ZADD_IN_INCR) != 0;
+  int nx = (in_flags & ZADD_IN_NX) != 0;
+  int xx = (in_flags & ZADD_IN_XX) != 0;
+  int gt = (in_flags & ZADD_IN_GT) != 0;
+  int lt = (in_flags & ZADD_IN_LT) != 0;
```
[the remark this answers is not in the document]
Comment (romange) [reply src/core/sorted_map.cc:91]: ok, but again - i just copied the code.

File: src/core/sorted_map.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (chakaz) [reply src/core/sorted_map.h:42]: I'd return a `unique_ptr` rather than document it, it's self documenting and safer. It also won't have any performance implications (just call `.release()` from the callsite)

File: src/core/sorted_map.h [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+ public:
+  using ScoredMember = std::pair<std::string, double>;
+  using ScoredArray = std::vector<ScoredMember>;
+
+  SortedMap();
+  SortedMap(const SortedMap&) = delete;
+  SortedMap& operator=(const SortedMap&) = delete;
+
+  SortedMap(SortedMap&& other) noexcept = default;
+  SortedMap& operator=(SortedMap&& other) noexcept = default;
```
Comment (chakaz) [inline src/core/sorted_map.h:37]: Actually I don't think that the default move operations are a good fit here, as they won't `null`ify the members of the moved-from object.
If this class, like you implied, is only ever used on the heap then you can `delete` those as well. If you do think it might be useful to have these movable, then I think you'll have to implement these operations yourself, and make sure that in case the members are `nullptr` the d'tor works as expected.

Comment (romange) [reply src/core/sorted_map.h:38]: good catch

**Review by chakaz — APPROVED** [review]
